src/server/middleware/redirects/index.js

```diff
@@ -20,6 +20,17 @@ module.exports = (req, res, next) => {
     dev: !isProduction
   })
 
+  // Disable trailing slash redirect for development mode.
+  // Because it leads to infinite loops as Gatsby in dev mode redirects to urls with trailing slashes
+  if (
+    !isProduction &&
+    location &&
+    location.startsWith('/') &&
+    parsedUrl.pathname === `${location}/`
```

Comment:
pathname === `${location}/` ?

so does it mean that in dev mode we do have `/`s?

Reply:
Yes. In dev mode we have trailing slashes and I'm not sure we can get rid of them at all... That's why we need extra handling in some another places.